flycheck.el

```diff
@@ -5440,6 +5440,75 @@ about TSLint."
              (and (not (string-empty-p output))
                   (json-read-from-string output)))))
 
+(defun flycheck-parse-rust (output checker buffer)
+  "Parse rust errors from JSON OUTPUT and return a list of `flycheck-error'.
+
+CHECKER and BUFFER denote the CHECKER that returned OUTPUT and
+the BUFFER that was checked respectively."
+  (let* ((json-array-type 'list)
+         (json-false nil)
+         ;; Cargo will output lines that are not the JSON errors we want, we
+         ;; skip them
+         (json-lines (seq-filter (lambda (line)
+                                   (string-match-p "^\{" line))
+                                 (split-string output "\n")))
+         (messages (seq-map #'json-read-from-string json-lines))
+         (errors))
+    ;; The JSON format of rust errors is described there:
+    ;; https://github.com/rust-lang/rust/blob/master/src/libsyntax/json.rs#L67-L139
+    ;; The gist of it is that each JSON line can contain one more spans.  Each
+    ;; span points to the root cause of error in the source text (the primary
+    ;; span), or to related causes.
+    (dolist (message messages (nreverse errors))
+      (let-alist message
+        (let ((error-message .message)
+              ;; Assumption: .level is "error" or "warning"
```

Comment:
Please use `pcase` to explicitly turn the strings into Flycheck levels and signal an explicit error or add an explicit default if an unknown level occurs.